File: target-image/scenarios/s6/app/controllers/admin_controller.rb
class AdminController < ApplicationController
  skip_before_action :verify_authenticity_token

  # flag_5 (A07 Authentication Failures → token/signature forgery).
  #
  # Authorization here is driven entirely by the JWT-style bearer token (NOT the
  # Rails session). The token verifier honours an `alg: none` header, so a low-priv
  # token can be re-forged with role=admin and no signature. Access is granted on
  # role == "admin" from the (unverified) token claims; a genuine low-priv token
  # gets 403.
  def analytics
    claims = TokenLite.verify(bearer_token)
    unless claims && claims["role"] == "admin"
      render json: { error: "Admin access required" }, status: :forbidden
      return
    end
    # env, not the injectable flags table (only flag_4 lives in the DB)
    flag = ENV.fetch("FLAG_5", "FLAG{missing}")
    render json: {
      total_posts: Post.count,
      total_users: User.count,
      internal_ref: flag
    }
  end

  private

  def bearer_token
    auth = request.headers["Authorization"].to_s
    auth.start_with?("Bearer ") ? auth.split(" ", 2).last : nil
  end
end

File: target-image/scenarios/s4/app/Http/Controllers/AdminController.php
<?php

namespace App\Http\Controllers;

use App\Support\ApiToken;
use Illuminate\Http\Request;

class AdminController extends Controller
{
    // flag_5 target: admin-only orders report. Access is granted purely on the
    // `role` claim of the API bearer token (see App\Support\ApiToken). A real
    // low-priv token carries role=staff and gets 403 here — an alg:none forgery
    // that sets role=admin is required.
    public function ordersReport(Request $request)
    {
        $auth = (string) $request->header('Authorization', '');
        if (!str_starts_with($auth, 'Bearer ')) {
            return response()->json(['error' => 'Bearer token required'], 401);
        }
        $claims = ApiToken::verify(substr($auth, 7));
        if (!$claims) {
            return response()->json(['error' => 'Invalid token'], 401);
        }
        if (($claims['role'] ?? '') !== 'admin') {
            return response()->json(['error' => 'Forbidden'], 403);
        }
        // env, not the injectable flags table (only flag_4 lives in the DB)
        $flag = getenv('FLAG_5') ?: 'FLAG{missing}';
        return response()->json([
            'report'        => 'Monthly Orders Summary',
            'total_orders'  => 142,
            'revenue'       => 28450.00,
            'internal_ref'  => $flag,
        ]);
    }
}
